flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,211 @@
+        /* Tuple2 : <sampleVector, label> */
+        DataStream<Tuple2<DenseVector, Double>> inputData =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new MapFunction<Row, Tuple2<DenseVector, Double>>() {
+                                    @Override
+                                    public Tuple2<DenseVector, Double> map(Row value) {
+                                        Double label = (Double) value.getField(getLabelCol());
+                                        DenseVector feature =
+                                                (DenseVector) value.getField(getFeaturesCol());
+                                        return Tuple2.of(feature, label);
+                                    }
+                                });
+        DataStream<KnnModelData> distributedModelData = prepareModelData(inputData);
+        DataStream<KnnModelData> modelData = mergeModelData(distributedModelData);
+        KnnModel model = new KnnModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return this.params;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static Knn load(StreamExecutionEnvironment env, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Prepares distributed knn model data. Constructs the sample matrix and computes norm of
+     * features.
+     *
+     * @param inputData Input feature data with label.
+     * @return Distributed knn model.
+     */
+    private static DataStream<KnnModelData> prepareModelData(
+            DataStream<Tuple2<DenseVector, Double>> inputData) {
+        return DataStreamUtils.mapPartition(
+                inputData,
+                new RichMapPartitionFunction<Tuple2<DenseVector, Double>, KnnModelData>() {
+                    @Override
+                    public void mapPartition(
+                            Iterable<Tuple2<DenseVector, Double>> values,
+                            Collector<KnnModelData> out) {
+                        List<Tuple2<DenseVector, Double>> dataPoints = new ArrayList<>(0);
+                        for (Tuple2<DenseVector, Double> tuple2 : values) {
+                            dataPoints.add(tuple2);
+                        }
+                        int featureDim = dataPoints.get(0).f0.size();
+                        DenseMatrix packedFeatures = new DenseMatrix(featureDim, dataPoints.size());
+                        DenseVector labels = new DenseVector(dataPoints.size());
+                        for (int i = 0; i < dataPoints.size(); ++i) {
+                            Tuple2<DenseVector, Double> tuple2 = dataPoints.get(i);
```

Comment:
nits: Could we rename `tuple2` to `dataPoint`? This is more consistent with `dataPoints`.

Reply:
OK